clustered/server/service/src/main/java/org/ehcache/clustered/server/state/EhcacheStateServiceProvider.java

```diff
@@ -61,6 +62,11 @@ public void addStateTo(StateDumpCollector dump) {
 
   @Override
   public boolean initialize(ServiceProviderConfiguration configuration, PlatformConfiguration platformConfiguration) {
+    if (!platformConfiguration.getExtendedConfiguration(TopologyService.class).iterator().next().isActivated()) {
```

Comment:
What are the rules on which services should and shouldn't get activated? If the rules are fixed and could be evaluated by Voltron itself couldn't we just have it not call initialize rather than having providers have to know this contract?

Reply:
There is no global rule. This is per service and extension. I.e, there are some service that need to be there because they are exposing some mbeans for diagnostic port. Security needs to be there also.

Reply:
FYI, I wanted to have a better / cleaner API (I proposed something with just an annotation added to the service) but it requires a sort of hack by taking control over the `Server` impl with  `Server.getServer()` / `Server.setServer()` to filter the loading of builtin services and configs.

Reply:
3rd point to consider: the service loading is a bit messy. Right now, core controls the loading of builtin services without giving DC a chance to veto. But DC controls the loading of extended configurations and service provider configurations (those creating service providers at the end). So there is IMO a responsibility issue here and until this gets correctly solved I don't see how we can have a clean solution.

A fix would be to let DC provide through the `StartupConfiguration` the list of builtin services, exactly like DC is doing for the other services right now, and not have core load them.

@myronkscott : FYI discussion above ^^ :-)

Reply:
If we're not going to fix this properly by separating core services (like those that support the config system) from the rest of the services, then wouldn't it just be simpler to enumerate the 5 special services somewhere and not force every service implementation to know about the wart?

Reply:
I’m curious why you care these services are loaded when unconfigured.  It doesn’t have consequences.  We should just fix persistence so that it does the correct action when presented with zero data-dirs.

Reply:
> If we're not going to fix this properly by separating core services (like those that support the config system) from the rest of the services, then wouldn't it just be simpler to enumerate the 5 special services somewhere 

This is exactly similar to what I wanted to do with these enumerations: having a way to define exceptionally some non default rules: i.e. `OnlyWhen({CONFIGURED, ACTIVATED})`. The default being `OnlyWhen(ACTIVATED)`. This change would remove 80% of the changes required. But it also require DC to hook in the `ServerEnv` to control how the built in serviced are loaded.

I don't like enumeration because it is just an ugly workaround that is worse than what I have proposed and this list is also spread amongst EE and OSS... Besides, having such list of built in services to load (or not load) would require a core update if core is still loading itself the builtin services. So ion that case, why no just do things correctly and have core not loaded these services and DC instead and do the annotation filtering.

> and not force every service implementation to know about the wart?

they already know about it ;-) the dependency is there.

Reply:
The wart is about the fact that dynamic config shouldn't be a service. It needs to be more tightly coupled (or maybe just differently coupled) to the core code. Services shouldn't need to know about any of that.

Reply:
This, and I suspect a huge number of other both discovered and undiscovered warts are all symptomatic of the underlying issue with the way that dynamic configuration integrates with core. I don't see the value in band-aid fixing these symptoms, especially if those band aids are just causing the warts to metastasize like this does. We either need to isolate the warts and not allow them to spread (aka ignore them), or we need to remove them by fixing the root problems.

Reply:
I agree with all that. I would like an approach where we do not touch the services excepts those we specifically want even in diagnostic mode. We can discuss that with @myronkscott later :-) I will ping you guys.

Reply:
> I’m curious why you care these services are loaded when unconfigured. It doesn’t have consequences. We should just fix persistence so that it does the correct action when presented with zero data-dirs.

Base on what we said yesterday, when in diagnostic mode, no extension are loaded. So no data dirs and no offheap. Most of the services right now are relying on plugins to be there (i.e. FRS).

What I would like to avoid also is going to each service and change their code to make them optional if they do not see a plugin because this is wrong: it would imply we can start an activated EE server without any data dirs no persistence, or ehcache without offheap resources. Some are invalid states and some are states that for now we didn't want to have (like no persistence in EE).

That's why I think it would be far easier to let DC control the loading of all the services and have a way to explicitly (annotations ?) mark the ones that should not be filtered out (by default) when running in diagnostic mode.